FirebaseCore/Sources/FIRFirebaseUserAgent.m

```diff
@@ -0,0 +1,150 @@
+            [components addObject:[NSString stringWithFormat:@"%@/%@", name, value]];
+          }];
+      [components sortUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
+      _firebaseUserAgent = [components componentsJoinedByString:@" "];
+    }
+    return _firebaseUserAgent;
+  }
+}
+
+- (void)setValue:(NSString *)value forComponent:(NSString *)componentName {
+  @synchronized(self) {
+    self.valuesByComponent[componentName] = value;
+    // Reset cached user agent string.
+    _firebaseUserAgent = nil;
+  }
+}
+
+- (void)reset {
+  @synchronized(self) {
+    // Reset components.
+    _valuesByComponent = [[[self class] environmentComponents] mutableCopy];
+    // Reset cached user agent string.
+    _firebaseUserAgent = nil;
+  }
+}
+
+#pragma mark - Environment components
+
++ (NSDictionary<NSString *, NSString *> *)environmentComponents {
+  NSMutableDictionary<NSString *, NSString *> *components = [NSMutableDictionary dictionary];
+
+  NSDictionary<NSString *, id> *info = [[NSBundle mainBundle] infoDictionary];
+  NSString *xcodeVersion = info[@"DTXcodeBuild"];
+  NSString *sdkVersion = info[@"DTSDKBuild"];
+
+  NSString *swiftFlagValue = [self hasSwiftRuntime] ? @"true" : @"false";
+  NSString *isFromAppstoreFlagValue = [GULAppEnvironmentUtil isFromAppStore] ? @"true" : @"false";
+
+  components[@"apple-platform"] = [self applePlatform];
+  components[@"apple-sdk"] = sdkVersion;
+  components[@"appstore"] = isFromAppstoreFlagValue;
+  components[@"deploy"] = [self deploymentType];
+  components[@"device"] = [GULAppEnvironmentUtil deviceModel];
+  components[@"os-version"] = [GULAppEnvironmentUtil systemVersion];
+  components[@"swift"] = swiftFlagValue;
+  components[@"xcode"] = xcodeVersion;
+
+  return [components copy];
+}
+
++ (BOOL)hasSwiftRuntime {
+  // The class
+  // [Swift._SwiftObject](https://github.com/apple/swift/blob/5eac3e2818eb340b11232aff83edfbd1c307fa03/stdlib/public/runtime/SwiftObject.h#L35)
+  // is a part of Swift runtime, so it should be present if Swift runtime is available.
+
+  BOOL hasSwiftRuntime =
+      objc_lookUpClass("Swift._SwiftObject") != nil ||
+      // Swift object class name before
+      // https://github.com/apple/swift/commit/9637b4a6e11ddca72f5f6dbe528efc7c92f14d01
+      objc_getClass("_TtCs12_SwiftObject") != nil;
+
+  return hasSwiftRuntime;
+}
+
++ (NSString *)applePlatform {
+  NSString *applePlatform = @"unknown";
+
+  // When a Catalyst app is run on macOS then both `TARGET_OS_MACCATALYST` and `TARGET_OS_IOS` are
+  // `true`, which means the condition list is order-sensitive.
+#if TARGET_OS_MACCATALYST
+  applePlatform = @"maccatalyst";
+#elif TARGET_OS_IOS
+  applePlatform = @"ios";
+#elif TARGET_OS_TV
+  applePlatform = @"tvos";
+#elif TARGET_OS_OSX
+  applePlatform = @"macos";
+#elif TARGET_OS_WATCH
+  applePlatform = @"watchos";
+#endif
+
+  return applePlatform;
+}
+
++ (NSString *)deploymentType {
```

Comment:
How will the corresponding code in CoreDiagnostics be deleted?

Reply:
I'll probably move the code down to GoogleUtilities so Core and CoreDiagnostics can use it from there.